# Problem Statement:
- Based on the different factors like marketing expense, production expense etc. predict the collection the movie can make once released.
- Collection of the movies is the target
- All other features are independent

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [2]:
data = pd.read_csv("Movie_regression.csv")
data.head()

Marketing expense  Production expense  Multiplex coverage     Budget  \
0            20.1264               59.62               0.462  36524.125   
1            20.5462               69.14               0.531  35668.655   
2            20.5458               69.14               0.531  39912.675   
3            20.6474               59.36               0.542  38873.890   
4            21.3810               59.36               0.542  39701.585   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0         138.7               7.825                8.095            7.910   
1         152.4               7.505                7.650            7.440   
2         134.6               7.485                7.570            7.495   
3         119.3               6.895                7.035            6.920   
4         127.7               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views 3D_available  Time_taken  \
0            7.995           7.94         527367          YES      109.60   
1            7.470           7.44         494055           NO      146.64   
2            7.515           7.44         547051           NO      147.88   
3            7.020           8.26         516279          YES      185.36   
4            7.070           8.26         531448           NO      176.48   

   Twitter_hastags     Genre  Avg_age_actors  Num_multiplex  Collection  
0          223.840  Thriller              23            494       48000  
1          243.456     Drama              42            462       43200  
2         2022.400    Comedy              38            458       69400  
3          225.344     Drama              45            472       66800  
4          225.792     Drama              55            395       72400

In [3]:
data.shape

(506, 18)

# Observations:
- The dataset contains 506 rows and 18 column
- We have a data of 506 movies

In [4]:
data.isnull().sum()

Marketing expense       0
Production expense      0
Multiplex coverage      0
Budget                  0
Movie_length            0
Lead_ Actor_Rating      0
Lead_Actress_rating     0
Director_rating         0
Producer_rating         0
Critic_rating           0
Trailer_views           0
3D_available            0
Time_taken             12
Twitter_hastags         0
Genre                   0
Avg_age_actors          0
Num_multiplex           0
Collection              0
dtype: int64

In [5]:
from sklearn.impute import SimpleImputer
si = SimpleImputer(strategy = "median")
si

SimpleImputer(strategy='median')

In [6]:
data["Time_taken"] = si.fit_transform(data[["Time_taken"]])
data.head()

Marketing expense  Production expense  Multiplex coverage     Budget  \
0            20.1264               59.62               0.462  36524.125   
1            20.5462               69.14               0.531  35668.655   
2            20.5458               69.14               0.531  39912.675   
3            20.6474               59.36               0.542  38873.890   
4            21.3810               59.36               0.542  39701.585   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0         138.7               7.825                8.095            7.910   
1         152.4               7.505                7.650            7.440   
2         134.6               7.485                7.570            7.495   
3         119.3               6.895                7.035            6.920   
4         127.7               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views 3D_available  Time_taken  \
0            7.995           7.94         527367          YES      109.60   
1            7.470           7.44         494055           NO      146.64   
2            7.515           7.44         547051           NO      147.88   
3            7.020           8.26         516279          YES      185.36   
4            7.070           8.26         531448           NO      176.48   

   Twitter_hastags     Genre  Avg_age_actors  Num_multiplex  Collection  
0          223.840  Thriller              23            494       48000  
1          243.456     Drama              42            462       43200  
2         2022.400    Comedy              38            458       69400  
3          225.344     Drama              45            472       66800  
4          225.792     Drama              55            395       72400

In [7]:
data.isnull().sum()

Marketing expense      0
Production expense     0
Multiplex coverage     0
Budget                 0
Movie_length           0
Lead_ Actor_Rating     0
Lead_Actress_rating    0
Director_rating        0
Producer_rating        0
Critic_rating          0
Trailer_views          0
3D_available           0
Time_taken             0
Twitter_hastags        0
Genre                  0
Avg_age_actors         0
Num_multiplex          0
Collection             0
dtype: int64

# Seperate Numerical and Categorical columns

In [8]:
num_cols = data.select_dtypes(["int","float"])
cat_cols = data.select_dtypes("object")

In [9]:
num_cols.columns

Index(['Marketing expense', 'Production expense', 'Multiplex coverage',
       'Budget', 'Movie_length', 'Lead_ Actor_Rating', 'Lead_Actress_rating',
       'Director_rating', 'Producer_rating', 'Critic_rating', 'Trailer_views',
       'Time_taken', 'Twitter_hastags', 'Avg_age_actors', 'Num_multiplex',
       'Collection'],
      dtype='object')

In [10]:
data.head()

Marketing expense  Production expense  Multiplex coverage     Budget  \
0            20.1264               59.62               0.462  36524.125   
1            20.5462               69.14               0.531  35668.655   
2            20.5458               69.14               0.531  39912.675   
3            20.6474               59.36               0.542  38873.890   
4            21.3810               59.36               0.542  39701.585   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0         138.7               7.825                8.095            7.910   
1         152.4               7.505                7.650            7.440   
2         134.6               7.485                7.570            7.495   
3         119.3               6.895                7.035            6.920   
4         127.7               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views 3D_available  Time_taken  \
0            7.995           7.94         527367          YES      109.60   
1            7.470           7.44         494055           NO      146.64   
2            7.515           7.44         547051           NO      147.88   
3            7.020           8.26         516279          YES      185.36   
4            7.070           8.26         531448           NO      176.48   

   Twitter_hastags     Genre  Avg_age_actors  Num_multiplex  Collection  
0          223.840  Thriller              23            494       48000  
1          243.456     Drama              42            462       43200  
2         2022.400    Comedy              38            458       69400  
3          225.344     Drama              45            472       66800  
4          225.792     Drama              55            395       72400

In [11]:
cat_cols.columns

Index(['3D_available', 'Genre'], dtype='object')

# Observation:
- We wont standardize Lead_ActorRating, Lead_Actress_Rating, Director_rating, Producer_rating, Critic_rating. The reason is the rating will be in the range of 0 - 10
- We wont standardize the Collection as well as it is the target.

In [14]:
num_cols = num_cols.drop(columns = ["Lead_Actress_rating", "Lead_ Actor_Rating", "Director_rating", "Producer_rating", "Critic_rating", "Collection"])
num_cols.head()

Marketing expense  Production expense  Multiplex coverage     Budget  \
0            20.1264               59.62               0.462  36524.125   
1            20.5462               69.14               0.531  35668.655   
2            20.5458               69.14               0.531  39912.675   
3            20.6474               59.36               0.542  38873.890   
4            21.3810               59.36               0.542  39701.585   

   Movie_length  Trailer_views  Time_taken  Twitter_hastags  Avg_age_actors  \
0         138.7         527367      109.60          223.840              23   
1         152.4         494055      146.64          243.456              42   
2         134.6         547051      147.88         2022.400              38   
3         119.3         516279      185.36          225.344              45   
4         127.7         531448      176.48          225.792              55   

   Num_multiplex  
0            494  
1            462  
2            458  
3            472  
4            395

In [20]:
num_columns = num_cols.columns.to_list()
num_columns

['Marketing expense',
 'Production expense',
 'Multiplex coverage',
 'Budget',
 'Movie_length',
 'Trailer_views',
 'Time_taken',
 'Twitter_hastags',
 'Avg_age_actors',
 'Num_multiplex']

In [15]:
from sklearn.preprocessing import StandardScaler
ss= StandardScaler()
ss

StandardScaler()

In [23]:
data[num_columns] = ss.fit_transform(data[num_columns])
data.head()

Marketing expense  Production expense  Multiplex coverage    Budget  \
0          -0.419782           -1.287909            0.144217  0.413672   
1          -0.417339           -0.593381            0.740262  0.194274   
2          -0.417342           -0.593381            0.740262  1.282714   
3          -0.416750           -1.306878            0.835284  1.016303   
4          -0.412482           -1.306878            0.835284  1.228577   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0     -0.120013               7.825                8.095            7.910   
1      0.367166               7.505                7.650            7.440   
2     -0.265812               7.485                7.570            7.495   
3     -0.809889               6.895                7.035            6.920   
4     -0.511180               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views 3D_available  Time_taken  \
0            7.995           7.94       1.125733          YES   -1.548999   
1            7.470           7.44       0.641896           NO   -0.350025   
2            7.515           7.44       1.411631           NO   -0.309887   
3            7.020           8.26       0.964686          YES    0.903330   
4            7.070           8.26       1.185007           NO    0.615887   

   Twitter_hastags     Genre  Avg_age_actors  Num_multiplex  Collection  
0        -0.353398  Thriller       -1.294408      -0.480510       48000  
1        -0.166000     Drama        0.225431      -0.781749       43200  
2        16.828839    Comedy       -0.094535      -0.819404       69400  
3        -0.339029     Drama        0.465405      -0.687612       66800  
4        -0.334750     Drama        1.265320      -1.412470       72400

# Feature Encoding

In [27]:
dic = {"YES" : 1, "NO" : 0}
data['3D_available'] = data['3D_available'].replace(dic)
data.head()

Marketing expense  Production expense  Multiplex coverage    Budget  \
0          -0.419782           -1.287909            0.144217  0.413672   
1          -0.417339           -0.593381            0.740262  0.194274   
2          -0.417342           -0.593381            0.740262  1.282714   
3          -0.416750           -1.306878            0.835284  1.016303   
4          -0.412482           -1.306878            0.835284  1.228577   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0     -0.120013               7.825                8.095            7.910   
1      0.367166               7.505                7.650            7.440   
2     -0.265812               7.485                7.570            7.495   
3     -0.809889               6.895                7.035            6.920   
4     -0.511180               6.920                7.070            6.815   

   Producer_rating  Critic_rating  Trailer_views  3D_available  Time_taken  \
0            7.995           7.94       1.125733             1   -1.548999   
1            7.470           7.44       0.641896             0   -0.350025   
2            7.515           7.44       1.411631             0   -0.309887   
3            7.020           8.26       0.964686             1    0.903330   
4            7.070           8.26       1.185007             0    0.615887   

   Twitter_hastags     Genre  Avg_age_actors  Num_multiplex  Collection  
0        -0.353398  Thriller       -1.294408      -0.480510       48000  
1        -0.166000     Drama        0.225431      -0.781749       43200  
2        16.828839    Comedy       -0.094535      -0.819404       69400  
3        -0.339029     Drama        0.465405      -0.687612       66800  
4        -0.334750     Drama        1.265320      -1.412470       72400

In [29]:
from sklearn.preprocessing import OneHotEncoder
ohe = OneHotEncoder()
ohe

OneHotEncoder()

In [33]:
data["Genre"].unique()

array(['Thriller', 'Drama', 'Comedy', 'Action'], dtype=object)

In [31]:
genre_ohe = ohe.fit_transform(data[["Genre"]])
genre_ohe

<506x4 sparse matrix of type '<class 'numpy.float64'>'
	with 506 stored elements in Compressed Sparse Row format>

In [32]:
genre_ohe.toarray()

array([[0., 0., 0., 1.],
       [0., 0., 1., 0.],
       [0., 1., 0., 0.],
       ...,
       [0., 1., 0., 0.],
       [0., 1., 0., 0.],
       [0., 1., 0., 0.]])

In [34]:
genre_ohe = pd.DataFrame(genre_ohe.toarray(), columns = ["Action", "Comedy", "Drama", "Thriller"])
genre_ohe

Action  Comedy  Drama  Thriller
0       0.0     0.0    0.0       1.0
1       0.0     0.0    1.0       0.0
2       0.0     1.0    0.0       0.0
3       0.0     0.0    1.0       0.0
4       0.0     0.0    1.0       0.0
..      ...     ...    ...       ...
501     1.0     0.0    0.0       0.0
502     1.0     0.0    0.0       0.0
503     0.0     1.0    0.0       0.0
504     0.0     1.0    0.0       0.0
505     0.0     1.0    0.0       0.0

[506 rows x 4 columns]

In [35]:
data = pd.concat([data, genre_ohe], axis = 1)
data.head()

Marketing expense  Production expense  Multiplex coverage    Budget  \
0          -0.419782           -1.287909            0.144217  0.413672   
1          -0.417339           -0.593381            0.740262  0.194274   
2          -0.417342           -0.593381            0.740262  1.282714   
3          -0.416750           -1.306878            0.835284  1.016303   
4          -0.412482           -1.306878            0.835284  1.228577   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0     -0.120013               7.825                8.095            7.910   
1      0.367166               7.505                7.650            7.440   
2     -0.265812               7.485                7.570            7.495   
3     -0.809889               6.895                7.035            6.920   
4     -0.511180               6.920                7.070            6.815   

   Producer_rating  Critic_rating  ...  Time_taken  Twitter_hastags     Genre  \
0            7.995           7.94  ...   -1.548999        -0.353398  Thriller   
1            7.470           7.44  ...   -0.350025        -0.166000     Drama   
2            7.515           7.44  ...   -0.309887        16.828839    Comedy   
3            7.020           8.26  ...    0.903330        -0.339029     Drama   
4            7.070           8.26  ...    0.615887        -0.334750     Drama   

   Avg_age_actors Num_multiplex  Collection  Action  Comedy  Drama  Thriller  
0       -1.294408     -0.480510       48000     0.0     0.0    0.0       1.0  
1        0.225431     -0.781749       43200     0.0     0.0    1.0       0.0  
2       -0.094535     -0.819404       69400     0.0     1.0    0.0       0.0  
3        0.465405     -0.687612       66800     0.0     0.0    1.0       0.0  
4        1.265320     -1.412470       72400     0.0     0.0    1.0       0.0  

[5 rows x 22 columns]

In [36]:
data = data.drop(columns = "Genre")
data.head()

Marketing expense  Production expense  Multiplex coverage    Budget  \
0          -0.419782           -1.287909            0.144217  0.413672   
1          -0.417339           -0.593381            0.740262  0.194274   
2          -0.417342           -0.593381            0.740262  1.282714   
3          -0.416750           -1.306878            0.835284  1.016303   
4          -0.412482           -1.306878            0.835284  1.228577   

   Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
0     -0.120013               7.825                8.095            7.910   
1      0.367166               7.505                7.650            7.440   
2     -0.265812               7.485                7.570            7.495   
3     -0.809889               6.895                7.035            6.920   
4     -0.511180               6.920                7.070            6.815   

   Producer_rating  Critic_rating  ...  3D_available  Time_taken  \
0            7.995           7.94  ...             1   -1.548999   
1            7.470           7.44  ...             0   -0.350025   
2            7.515           7.44  ...             0   -0.309887   
3            7.020           8.26  ...             1    0.903330   
4            7.070           8.26  ...             0    0.615887   

   Twitter_hastags  Avg_age_actors  Num_multiplex  Collection  Action  Comedy  \
0        -0.353398       -1.294408      -0.480510       48000     0.0     0.0   
1        -0.166000        0.225431      -0.781749       43200     0.0     0.0   
2        16.828839       -0.094535      -0.819404       69400     0.0     1.0   
3        -0.339029        0.465405      -0.687612       66800     0.0     0.0   
4        -0.334750        1.265320      -1.412470       72400     0.0     0.0   

   Drama  Thriller  
0    0.0       1.0  
1    1.0       0.0  
2    0.0       0.0  
3    1.0       0.0  
4    1.0       0.0  

[5 rows x 21 columns]

# Seperate X and y

In [37]:
X = data.drop(columns = "Collection")
y = data["Collection"]

In [38]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

# Apply DecisionTreeRegressor on train set

In [39]:
from sklearn.tree import DecisionTreeRegressor
dtr = DecisionTreeRegressor()
dtr

DecisionTreeRegressor()

In [41]:
dtr.fit(X_train, y_train)

DecisionTreeRegressor()

# Performing Predictions on X_test

In [43]:
y_pred = dtr.predict(X_test)
y_pred

array([ 37200.,  58800.,  16200.,  48200.,  40000.,  44600.,  35000.,
        43000.,  40800.,  41400.,  37600.,  32000.,  24600.,  47800.,
        36400.,  74400.,  40200.,  24200., 100000.,  23400.,  49200.,
        44400.,  27000.,  60200.,  17600.,  27600.,  42000.,  29800.,
        34800.,  46600.,  37600.,  33000.,  35600.,  45400.,  35400.,
        30400.,  74600.,  41200.,  46400.,  46400.,  44800.,  74400.,
        96600.,  36400.,  49200.,  27600.,  38600.,  47800.,  36600.,
        59600.,  39200.,  66200.,  33200.,  45600., 100000.,  40200.,
        30400.,  59800.,  45800.,  35000.,  45600.,  69200.,  53200.,
        32200.,  49000.,  31200.,  29200.,  44600.,  59600.,  40000.,
        43600.,  45600.,  14800.,  36400.,  45800.,  14400.,  37600.,
       100000.,  34200.,  16200.,  43600.,  26600.,  39200.,  23600.,
        38400.,  47600.,  26800.,  45800.,  48200.,  36000.,  41200.,
        21000.,  29000.,  35000.,  38200.,  39800., 100000.,  26800.,
        33400.,  238

In [44]:
X_test

Marketing expense  Production expense  Multiplex coverage    Budget  \
173          -0.409837           -1.034027            0.386091  0.187151   
274          -0.413949           -0.689681            0.930305  0.674384   
491          -0.408212            2.422565           -0.469104 -0.429726   
72           -0.409853           -0.047680            1.224009 -0.312904   
452           0.171842            1.015999           -1.367490  0.017617   
..                 ...                 ...                 ...       ...   
412           1.768579            1.015999           -0.365444 -2.360139   
436           1.257678            1.015999           -1.600725  0.251261   
411           1.214608            1.015999           -0.365444  0.530494   
86           -0.414480           -0.969827            0.913029 -0.384137   
75           -0.409448            0.247057            1.016689  0.001946   

     Movie_length  Lead_ Actor_Rating  Lead_Actress_rating  Director_rating  \
173      0.552081               8.570                8.715            8.615   
274     -1.268620               7.900                8.010            7.810   
491      1.074822               9.020                9.155            8.970   
72      -2.161190               7.270                7.520            7.195   
452      0.825898               8.760                8.870            8.815   
..            ...                 ...                  ...              ...   
412      1.117494               9.115                9.230            9.110   
436      0.879239               8.840                9.090            8.945   
411      1.117494               9.220                9.295            9.170   
86      -0.834781               7.690                7.920            7.755   
75      -0.838337               7.585                7.750            7.585   

     Producer_rating  Critic_rating  Trailer_views  3D_available  Time_taken  \
173            8.805           7.68       0.255576             1    1.294356   
274            8.120           8.48       0.876494             0    1.796734   
491            9.120           6.98      -0.974918             0   -0.518348   
72             7.440           7.16       0.732455             0   -1.305579   
452            8.820           7.96      -0.555438             0   -0.744936   
..               ...            ...            ...           ...         ...   
412            9.435           7.96      -3.139316             0    0.314201   
436            9.125           6.96      -0.651779             1   -1.152794   
411            9.260           7.96      -1.406060             0    1.146750   
86             7.780           7.30      -0.190382             1   -1.244724   
75             8.070           8.26       0.679746             0   -1.194227   

     Twitter_hastags  Avg_age_actors  Num_multiplex  Action  Comedy  Drama  \
173        -0.545076        1.585286      -0.047478     0.0     1.0    0.0   
274        -0.149491       -0.734468      -1.120644     0.0     0.0    0.0   
491         0.203905        0.625388       0.169038     0.0     1.0    0.0   
72         -0.546299       -1.534383      -1.732536     0.0     0.0    0.0   
452         0.398793       -1.454391       0.950378     0.0     1.0    0.0   
..               ...             ...            ...     ...     ...    ...   
412         0.401545        0.065448       0.394968     0.0     1.0    0.0   
436        -0.566475        1.425303       1.477547     0.0     0.0    0.0   
411         0.209408        0.385414       0.065487     0.0     1.0    0.0   
86         -0.355690        0.865363      -0.687612     0.0     0.0    0.0   
75         -0.548438       -0.894451      -1.534848     0.0     1.0    0.0   

     Thriller  
173       0.0  
274       1.0  
491       0.0  
72        1.0  
452       0.0  
..        ...  
412       0.0  
436       1.0  
411       0.0  
86        1.0  
75        0.0  

[102 rows x 20 columns]

# Checking Accuracy

In [45]:
from sklearn.metrics import r2_score
r2_score(y_pred, y_test)

0.8380569806683374

In [46]:
from sklearn.metrics import mean_squared_error
np.sqrt(mean_squared_error(y_test, y_pred))

7116.757358103313

# Observations:

- Error is indicating that there can be a difference of  7120  in the predicted collection values
- eg: the first predicted value is 37200, with the error, the range of the collection will be $ 30700  to 44320

In [47]:
from sklearn.model_selection import cross_val_score
cross_val_score(dtr, X, y, cv = 10)

array([ 0.30613309,  0.51719667, -1.80983134,  0.13246601,  0.84355635,
        0.48232422,  0.29263656,  0.16143169, -2.68769334, -0.91922228])

X_test = smaller values y_pred smaller
y_test = larger

y_pred - y_test = smaller value - larger value = -ve